### DISCUSSION TITLE
Phase separation in block copolymers

### DISCUSSION THREAD

**Discussion by rnpratoori** [idaholab/moose#27827] [Q&A Modules: Phase field] [answered]
Hello team,

I am new to MOOSE and C++. I want to simulate phase separation in block copolymers. For, that I want to change the split formulation to reflect these equations -
![image](https://github.com/idaholab/moose/assets/58669729/59966d6d-206d-4efe-836f-ec07eaa3de21)
Essentially, there is one extra term in the residual of chemical potential.

I have created an auxkernel based modifying SplitCHResBase.h. Here are the new *.h and *.C files I created - [SplitCHPhaseSep.zip](https://github.com/user-attachments/files/15742765/SplitCHPhaseSep.zip). Is this the right way to do it? Is there a better way?

When I compile with these files, I get the error -
`error: no match for 'operator+=' (operand types are 'libMesh::Real' {aka 'double'} and 'libMesh::boostcopy::enable_if_c<true, libMesh::TypeVector<double> >::type' {aka 'libMesh::TypeVector<double>'})` 
This error appears in `computeQpResidual`, `computeQpWJacobian` and `computeQpOffDiagJacobian`.

Can anyone help me identify what mistake I might be making?

Thanks in advance,
Raghu.

**Comment by GiudGiud** [comment, marked as answer]
Hello

```
error: no match for 'operator+=' (operand types are 'libMesh::Real' {aka 'double'} and 'libMesh::boostcopy::enable_if_c<true, libMesh::TypeVector<double> >::type' {aka 'libMesh::TypeVector<double>'})
```
means you are summing a vector and a double. You have to use a loop to do that or maybe you do not want this?
```
  residual += _mob[_qp] * _alpha[_qp] * _kappa[_qp] * (_c[_qp] - _bar_c) * _grad_test[_i][_qp]; //Second term
```
residual is a real. 
The other term is a real times a vector (gradient of the i-th test function at quadrature point qp)
maybe you mean to use `_test` instead?

**Reply by rnpratoori** [reply]
Hello @GiudGiud for your reply. What you are saying makes sense, but I am wondering why I am not getting the same error in the first term where I use 
```Real residual = _mob[_qp] * _grad_w[_qp] * _grad_test[_i][_qp]; //First term```

This is also the same expression for `computeQpResidual` of `SplitCHWResBase.h`. Hence I believe it should be `_grad_test` and not just `_test`. Can you clarify this for me?

**Reply by GiudGiud** [reply]
Because in that first term you have a grad_w vector term, that causes a dot product with grad_test and you end up with a real.

**Reply by rnpratoori** [reply]
This makes perfect sense and fixed the errors. Thank you.
But, using this in the input file gives a new error -
```Task add_aux_kernel is not registered to build Kernel derived objects```
I am unsure if I am using the AuxKernel I developed correctly. Here is the input file (only the relevant parts) -
```
[Variables]
    # difference in the volume fractions of the 2 phases
    [./u]
        order = FIRST
        family = LAGRANGE
        [./InitialCondition]
            type = RandomIC
            seed = 123
            min = -0.1
            max =  0.1
        [../]
    [../]
    # Chemical potential (J/mol)
    [./w]
        order = FIRST
        family = LAGRANGE
    [../]
[]
  
[AuxVariables]
    # polymer volume fraction
    [./pvf]
        # order = FIRST
        # family = LAGRANGE
    [../]
[]
  
[Kernels]
    [./w_dot]
        type = CoupledTimeDerivative
        variable = w
        v = u
    [../]
[]
  
[AuxKernels]
    # calculate polymer volume fraction from difference in volume fractions
    [./pvf]
        type = ParsedAux
        variable = pvf
        coupled_variables = 'u'
        expression = '(u+1)/2'
    [../]
    # adding nonlocal term to the energy
    [./coupled_res]
        type = SplitCHPhaseSep
        variable = w
        mob_name = M
        c = u
        alpha_name = alpha
        kappa_name = kappa
    [../]
[]
[Materials]
    # Units of M are m^2 mol / (J s)
    # Units of kappa are J m^2 / mol
    # consider adding a scaling factor in the future
    [./mat]
        type = GenericConstantMaterial
        prop_names  = 'M   kappa    alpha'
        prop_values = '1e-02    5e-03   1e-01'
                        # M*mum_m^2/eV_J
                        # kappa*eV_J*mu_m^2
    [../]
[]
```
Can you point to what I might be missing or any example problem that deals with a similar thing?

**Reply by GiudGiud** [reply]
You are adding a kernel in the auxkernel block. 

Kernels should go in the [Kernels] block . They are meant to contribute to the residual and Jacobian

Auxkernels are meant to set auxiliary variables.

**Reply by rnpratoori** [reply]
Thank you @GiudGiud, this worked.
I am trying to calculate the element average value of a variable in my Kernel. What is the best way to do it? I do not need to output it. 
1. Should I use something like `QBase::n_points()` and do a `sum/n_points`, or
2. use `ElementAverageValue` or `ElementIntegralPostprocessor` to calculate it

**Reply by GiudGiud** [reply]
Do you need an average over a single element?
Or an average over the entire mesh?

Are you using automatic differentiation? (AD)

> Should I use something like QBase::n_points() and do a sum/n_points, or

This would be for computing the average on each element. There are other options too

> use ElementAverageValue or ElementIntegralPostprocessor to calculate it

This would be for computing the average over a larger region of the mesh like a subdomain

**Reply by rnpratoori** [reply]
I am looking for an average for each element. I am not using AD - I defined the Jacobian that needs to be calculated. Can you tell me what options I have to calculate this? I need to use it in calculating the residual -
```residual += _mob[_qp] * _alpha[_qp] * _kappa[_qp] * (_c[_qp] - _bar_c) * _test[_i][_qp];```
`_bar_c` is the average value of `_c[_qp]` over an element.

**Reply by GiudGiud** [reply]
If you are not using AD, then you could create an auxiliary variable bar_c of type CONSTANT MONOMIAL.
Then use the ProjectionAux to project c onto bar_c on every linear iterations. This computes bar_c as the average of c on each element

**Reply by rnpratoori** [reply]
That is an interesting way! Can you tell me how to initialize the `_bar_c` variable now?
I know I have to use - `_bar_c(coupledValue()),`, but unsure of what should be the variable name to be used in coupledValue

**Reply by GiudGiud** [reply]
in the object, add:
```
params.addCoupledVar("bar_c", ...); 
```
to the validParams

then you can use 
```
...,
_bar_c(coupledValue("bar_c")),
```
in the initializer's list of the constructor of your object

**Reply by rnpratoori** [reply]
I have one more question. I want to calculate the total energy of the system. When it was simple CH equation, I could use `TotalFreeEnergy` to calculate it. Now, can I still use the same? I noticed that there is a `additional_free_energy` input parameter - is this for the $E_d$ term in the Basic Phase Field Equations? How can I use it with my system or is there a better way I can do it?

**Reply by GiudGiud** [reply]
I dont know. @tonkmr maybe you know?
